File: Problems/Algorithms/173. Binary Search Tree Iterator/BSTIterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {
    private Queue<Integer> queue;

    public BSTIterator(TreeNode root) {
        queue = new LinkedList<>();
        this.inorder(root);
    }
    
    public int next() {
        return this.queue.poll();
    }
    
    public boolean hasNext() {
        return !this.queue.isEmpty();
    }
    
    private void inorder(TreeNode node) {
        if (node == null) return;
        
        inorder(node.left);
        queue.add(node.val);
        inorder(node.right);
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: Problems/Algorithms/173. Binary Search Tree Iterator/BSTIterator.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class BSTIterator(root: TreeNode?) {
    
    private var stack = ArrayDeque<TreeNode?>()
    
    init {
        this.goLeft(root)
    }

    fun next(): Int {
        var curr: TreeNode? = this.stack.removeLast()
        
        if (curr?.right != null) {
            this.goLeft(curr?.right)
        }
        
        return curr?.`val`!!
    }

    fun hasNext(): Boolean {
        return !this.stack.isEmpty()
    }
    
    fun goLeft(node: TreeNode?): Unit {
        var curr: TreeNode? = node
        
        while (curr != null) {
            this.stack.add(curr)
            curr = curr.left
        }
    }

}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */
